src/Shared/FileUtilities.cs

```diff
@@ -361,37 +361,54 @@ internal static string MaybeAdjustFilePath(string value, string baseDirectory =
         {
             // Don't bother with arrays or properties or network paths, or those that
             // have no slashes.
-            if (NativeMethodsShared.IsWindows || string.IsNullOrEmpty(value) ||
-                value.StartsWith("$(") || value.StartsWith("@(") || value.StartsWith("\\\\") ||
-                value.IndexOfAny(Slashes) == -1)
+            if (NativeMethodsShared.IsWindows || string.IsNullOrEmpty(value)
+                || value.StartsWith("$(", PathComparison) || value.StartsWith("@(", PathComparison)
+                || value.StartsWith("\\\\", PathComparison))
             {
                 return value;
             }
 
             // For Unix-like systems, we may want to convert backslashes to slashes
-            string newValue = Regex.Replace(value, @"[\\/]+", "/");
+            string newValue = ConvertToUnixSlashes(value);
 
-            string quote = string.Empty;
             // Find the part of the name we want to check, that is remove quotes, if present
-            string checkValue = newValue;
-            if (newValue.Length > 2)
+            bool shouldAdjust = newValue.IndexOf('/') != -1 && LooksLikeUnixFilePath(RemoveQuotes(newValue), baseDirectory);
+            return shouldAdjust ? newValue : value;
+        }
+
+
+        private static string ConvertToUnixSlashes(string path)
+        {
+            if (path.IndexOf('\\') == -1)
             {
-                if (newValue.StartsWith("'"))
-                {
-                    if (newValue.EndsWith("'"))
-                    {
-                        checkValue = newValue.Substring(1, newValue.Length - 2);
-                        quote = "'";
-                    }
-                }
-                else if (newValue.StartsWith("\"") && newValue.EndsWith("\""))
-                {
-                    checkValue = newValue.Substring(1, newValue.Length - 2);
-                    quote = "\"";
+                return path;
+            }
+            var unixPath = new StringBuilder(path.Length);
```

Comment:
Can you use the pool? https://github.com/Microsoft/msbuild/blob/master/src/Shared/ReuseableStringBuilder.cs

If we had one that was say `MAX_PATH` long (😜) it could be reused an awful lot here, and avoid allocations.

Reply:
Awesome, I was digging around for a StringBuilderCache or similar and completely missed that

Reply:
Actually MSBuildTaskHost didn't like that at all thanks to ```Concurrent```. If there isn't a way around that I found we do have ```StringBuilderCache``` already ripped straight from mscorlib, although that's only per thread and always allocates a new one for long strings.

Reply:
Well, at least 360 > `MAX_PATH`. That seems like a reasonable option, better than using `Reusable` + a bunch of ifdefs for .NET 3.5 compat.